File: asmt3-eng-auction/chaincode-javascript/lib/engAuction.js
/*
 * 
 * SPDX-License-Identifier: Apache-2.0
 */

'use strict';

const { Contract } = require('fabric-contract-api');

class EnglishAuction extends Contract {

    async InitLedger(ctx) {

        const assets = [
            {
                ID: 'asset1',
                Owner: 'Test1',
                ReservedPrice: 300,
                BidIncrement: 50,
                BidDecrement: 10
            },
            
        ];
        const autionInfo = {
            ID: 'auctionInfo',
            BidValue: 0,
            BidderID: '',
            assetID: 'asset1',
            numOrgsSubmitted: 0,
            numBelowReserved: 0
        }
        const auctionSubmitters = [
            {
                ID: '1',
                MSP: 'none',
            },
            {
                ID: '2',
                MSP: 'none',
            },
            {
                ID: '3',
                MSP: 'none',
            },
        ]
       
        for (const asset of assets) {
            asset.docType = 'asset';
            await ctx.stub.putState(asset.ID, Buffer.from(JSON.stringify(asset)));
            console.info(`Asset ${asset.ID} initialized`);
        }
        for (const submitter of auctionSubmitters) {
            submitter.docType = 'submitters';
            await ctx.stub.putState(submitter.ID, Buffer.from(JSON.stringify(submitter)));
            console.info(`Submitter ${submitter.ID} initialized`);
        }
        await ctx.stub.putState(autionInfo.ID, Buffer.from(JSON.stringify(autionInfo)));
    }

    // CreateAsset issues a new asset to the world state with given details.
    async CreateAsset(ctx, id, owner, reservedPrice, bidIncr, bidDecr) {
        const asset = {
            ID: id,
            Owner: owner,
            ReservedPrice: reservedPrice,
            BidIncrement: bidIncr,
            BidDecrement: bidDecr,
        };
        return ctx.stub.putState(id, Buffer.from(JSON.stringify(asset)));
    }

    // ReadAsset returns the asset stored in the world state with given id.
    async ReadAsset(ctx, id) {
        const assetJSON = await ctx.stub.getState(id); // get the asset from chaincode state
        if (!assetJSON || assetJSON.length === 0) {
            throw new Error(`The asset ${id} does not exist`);
        }
        return assetJSON.toString();
    }

    // UpdateAsset updates an existing asset in the world state with provided parameters.
    async UpdateAsset(ctx, id, owner, reservedPrice, bidIncr, bidDecr) {
        const exists = await this.AssetExists(ctx, id);
        if (!exists) {
            throw new Error(`The asset ${id} does not exist`);
        }

        // overwriting original asset with new asset
        const updatedAsset = {
            ID: id,
            Owner: owner,
            ReservedPrice: reservedPrice,
            BidIncrement: bidIncr,
            BidDecrement: bidDecr,
        };
        return ctx.stub.putState(id, Buffer.from(JSON.stringify(updatedAsset)));
    }

    // DeleteAsset deletes an given asset from the world state.
    async DeleteAsset(ctx, id) {
        const exists = await this.AssetExists(ctx, id);
        if (!exists) {
            throw new Error(`The asset ${id} does not exist`);
        }
        return ctx.stub.deleteState(id);
    }

    // AssetExists returns true when asset with given ID exists in world state.
    async AssetExists(ctx, id) {
        const assetJSON = await ctx.stub.getState(id);
        return assetJSON && assetJSON.length > 0;
    }

    // TransferAsset updates the owner field of asset with given id in the world state.
    async TransferAsset(ctx, id, newOwner) {
        const assetString = await this.ReadAsset(ctx, id);
        const asset = JSON.parse(assetString);
        asset.Owner = newOwner;
        return ctx.stub.putState(id, Buffer.from(JSON.stringify(asset)));
    }

    // GetAllAssets returns all assets found in the world state.
    async GetAllAssets(ctx) {
        const allResults = [];
        // range query with empty string for startKey and endKey does an open-ended query of all assets in the chaincode namespace.
        const iterator = await ctx.stub.getStateByRange('asset1', 'auctionInfo');
        let result = await iterator.next();
        while (!result.done) {
            const strValue = Buffer.from(result.value.value.toString()).toString('utf8');
            let record;
            try {
                record = JSON.parse(strValue);
            } catch (err) {
                console.log(err);
                record = strValue;
            }
            allResults.push({ Key: result.value.key, Record: record });
            result = await iterator.next();
        }
        return JSON.stringify(allResults);
    }
    async SubmitBid(ctx, bidValue, assetId) {
        const assetString = await this.ReadAsset(ctx, assetId);
        const asset = JSON.parse(assetString);
        // console.log("asset1", asset);
        const auctionB = await ctx.stub.getState('auctionInfo'); 
        var auctionJSON = JSON.parse(auctionB.toString());
        // console.log("auctionJSON",auctionJSON);
        const submitterId = await ctx.clientIdentity.getMSPID();
        console.log("submitterId",submitterId);
        var i;
        var nullSubmitterIdx = [];
        var exists = false;
        var idx = 0;
        for (i = 1; i < 4; i++) {
            const submitterB = await ctx.stub.getState(i.toString());
            var submitter = JSON.parse(submitterB.toString());
            if (submitter.MSP == submitterId){
                exists = true;
                break;
            }else if (submitter.MSP == 'none'){
                nullSubmitterIdx[idx++] = i;
            }
        }
        if(exists == false){
            console.log("nullSubmitterIdx",nullSubmitterIdx )
            var id = 0;
            if (nullSubmitterIdx.length > 0){
                id = nullSubmitterIdx[0];
            }
            const submitter1 = {
                ID: id.toString(),
                MSP: submitterId
            };
            await ctx.stub.putState( submitter1.ID, Buffer.from(JSON.stringify(submitter1)));
            auctionJSON.numOrgsSubmitted++;
        }
        // console.log(auctionJSON);
        // console.log(bidValue, auctionJSON.BidValue, assetId, auctionJSON.assetID)
        if (auctionJSON.assetID == assetId) {
            if (asset.ReservedPrice <= bidValue){
                if (auctionJSON.BidValue < bidValue){
                    auctionJSON.BidValue = bidValue;
                    auctionJSON.BidderID = submitterId;       
                }         
            }else {
                console.log("bid Value less Than reserved Price");  
                auctionJSON.numBelowReserved++;
            }
        }
        if (auctionJSON.numBelowReserved >= 3){
            auctionJSON.numBelowReserved = 0;
            asset.ReservedPrice -= asset.BidDecrement;
        }
        //  console.log("auctionJSONend", auctionJSON);
         await ctx.stub.putState(assetId, Buffer.from(JSON.stringify(asset)));
         await ctx.stub.putState('auctionInfo', Buffer.from(JSON.stringify(auctionJSON)));
        
    }
    async DeclareWinner(ctx) {
        var winner = 'None'
        const auctionB = await ctx.stub.getState('auctionInfo'); 
        var auctionJSON = JSON.parse(auctionB.toString());
        console.log("auctionJSON",auctionJSON);
        if (auctionJSON.numOrgsSubmitted >= 3){
            console.log("Bidder ", auctionJSON.BidderID, " is winner of the Auction with value " ,  auctionJSON.BidValue);
            winner = auctionJSON.BidderID;
        }else {
            console.log("Aiuction not complete");
        }
        return winner;
    }


}

module.exports = EnglishAuction;

File: asmt3-eng-auction/chaincode-go/engAuction.go
package main

import (
	"encoding/json"
	"fmt"
	"log"
	"strconv"

	"github.com/hyperledger/fabric-contract-api-go/contractapi"
)

// SmartContract provides functions for managing an Asset
type SmartContract struct {
	contractapi.Contract
}

// Asset describes basic details of what makes up a simple asset
type Asset struct {
	ID            string `json:"ID"`
	Owner         string `json:"Owner"`
	ReservedPrice int    `json:"ReservedPrice"`
	BidIncrement  int    `json:"BidIncrement"`
	BidDecrement  int    `json:"BidDecrement"`
}

type auction struct {
	ID               string `json:"ID"`
	BidValue         int    `json:"BidValue"`
	BidderID         string `json:"BidderID"`
	AssetID          string `json:"AssetID"`
	NumOrgsSubmitted int    `json:"NumOrgsSubmitted"`
	NumBelowReserved int    `json:"NumBelowReserved"`
}

type identity struct {
	ID  string `json:"ID"`
	MSP string `json:"MSP"`
}

// InitLedger adds a base set of assets to the ledger
func (s *SmartContract) InitLedger(ctx contractapi.TransactionContextInterface) error {
	assets := []Asset{
		{
			ID:            "asset1",
			Owner:         "Test1",
			ReservedPrice: 20,
			BidIncrement:  50,
			BidDecrement:  10,
		},
	}

	auctionInfo := auction{
		ID:               "auctionInfo",
		BidValue:         0,
		BidderID:         "",
		AssetID:          "asset1",
		NumOrgsSubmitted: 0,
		NumBelowReserved: 0,
	}

	auctionSubmitters := []identity{
		{
			ID:  "1",
			MSP: "none",
		},
		{
			ID:  "2",
			MSP: "none",
		},
		{
			ID:  "3",
			MSP: "none",
		},
	}

	for _, asset := range assets {
		assetJSON, err := json.Marshal(asset)
		if err != nil {
			return err
		}

		err = ctx.GetStub().PutState(asset.ID, assetJSON)
		if err != nil {
			return fmt.Errorf("Failed to put to world state. %v", err)
		}
		// log.Printf("Asset %s initialized", asset.ID)
	}

	for _, submitter := range auctionSubmitters {
		submitterJSON, err := json.Marshal(submitter)
		if err != nil {
			return err
		}
		err = ctx.GetStub().PutState(submitter.ID, submitterJSON)
		if err != nil {
			return fmt.Errorf("Failed to put to world state. %v", err)
		}
		// log.Printf("Submitter %s initialized", submitter.ID)
	}

	auctionJSON, err := json.Marshal(auctionInfo)
	if err != nil {
		return err
	}
	err = ctx.GetStub().PutState(auctionInfo.ID, auctionJSON)
	return nil
}

// ReadAsset returns the asset stored in the world state with given id.
func (s *SmartContract) ReadAsset(ctx contractapi.TransactionContextInterface, id string) (*Asset, error) {
	assetPtr, err := ctx.GetStub().GetState(id)
	if err != nil {
		return nil, fmt.Errorf("failed to read from world state: %v", err)
	}
	if assetPtr == nil {
		return nil, fmt.Errorf("the asset %s does not exist", id)
	}

	var asset Asset
	err = json.Unmarshal(assetPtr, &asset)
	if err != nil {
		return nil, err
	}

	return &asset, nil
}

func (s *SmartContract) SubmitBid(ctx contractapi.TransactionContextInterface, bidValue int, assetID string) (*Asset, error) {
	assetString, err := s.ReadAsset(ctx, assetID)
	asset := assetString
	// asset, err := json.Marshal(assetString)
	// if err != nil {
	// 	return nil, err
	// }
	auctionPtr, err := ctx.GetStub().GetState("auctionInfo")
	if err != nil {
		return nil, fmt.Errorf("failed to read from world state: %v", err)
	}
	if auctionPtr == nil {
		return nil, fmt.Errorf("Auction Info absent")
	}
	var auctionJSON auction
	err = json.Unmarshal(auctionPtr, &auctionJSON)
	if err != nil {
		return nil, err
	}

	submitterID, err := ctx.GetClientIdentity().GetMSPID()
	if err != nil {
		return nil, fmt.Errorf("failed to read from world state: %v", err)
	}
	nullSubmitterIdx := [5]int{-1, -1, -1, -1, -1}
	var exists bool = false
	var idx int = 0

	for i := 1; i < 4; i++ {
		submitterPtr, err := ctx.GetStub().GetState(strconv.Itoa(i))
		if err != nil {
			return nil, fmt.Errorf("failed to read from world state: %v", err)
		}
		var submitter identity
		err = json.Unmarshal(submitterPtr, &submitter)
		if err != nil {
			return nil, err
		}
		// submitter := submitterJSON
		if submitter.MSP == submitterID {
			exists = true
			break
		} else if submitter.MSP == "none" {
			nullSubmitterIdx[idx] = i
			idx++
		}
	}

	if !exists {
		// log.Printf("Null Submitte Id!")
		id := 0
		if nullSubmitterIdx[0] != -1 {
			id = nullSubmitterIdx[0]
		}

		submitter1 := identity{
			ID:  strconv.Itoa(id),
			MSP: submitterID,
		}

		submitter1JSON, err := json.Marshal(submitter1)
		if err != nil {
			return nil, err
		}
		err = ctx.GetStub().PutState(submitter1.ID, submitter1JSON)
		auctionJSON.NumOrgsSubmitted++
	}

	if auctionJSON.AssetID == assetID {
		if asset.ReservedPrice <= bidValue {
			if auctionJSON.BidValue < bidValue {
				auctionJSON.BidValue = bidValue
				auctionJSON.BidderID = submitterID
			}
		} else {
			// log.Printf("bid Value less Than reserved Price")
			auctionJSON.NumBelowReserved++
		}
	}

	if auctionJSON.NumBelowReserved >= 3 {
		auctionJSON.NumBelowReserved = 0
		asset.ReservedPrice -= asset.BidDecrement
	}

	assetPtr, err := json.Marshal(asset)
	if err != nil {
		return nil, err
	}

	auctionPtr, err = json.Marshal(auctionJSON)
	if err != nil {
		return nil, err
	}

	ctx.GetStub().PutState(assetID, assetPtr)
	ctx.GetStub().PutState("auctionInfo", auctionPtr)
	return nil, nil
}

func (s *SmartContract) DeclareWinner(ctx contractapi.TransactionContextInterface) (*auction, error) {
	auctionPtr, err := ctx.GetStub().GetState("auctionInfo")

	var auctionJSON auction
	err = json.Unmarshal(auctionPtr, &auctionJSON)
	if err != nil {
		return nil, err
	}

	// if auctionJSON.NumOrgsSubmitted >= 3 {
	// 	log.Printf("Bidder %s wins the auction at Bid Value %d", auctionJSON.BidderID, auctionJSON.BidValue)
	// } else {
	// 	fmt.Print("Auction Incomplete")
	// }

	return &auctionJSON, nil
}

// start it all
func main() {
	assetChaincode, err := contractapi.NewChaincode(&SmartContract{})
	if err != nil {
		log.Panicf("Error creating asset-transfer-basic chaincode: %v", err)
	}

	if err := assetChaincode.Start(); err != nil {
		log.Panicf("Error starting asset-transfer-basic chaincode: %v", err)
	}
}
